### ISSUE TITLE
CPU LayerNorm normalized_shape=(1,) gives nonzero weight.grad although output is independent of weight

### ISSUE THREAD

**Issue by Robertboy18** [pytorch/pytorch#188372]
### Bug

For `torch.nn.functional.layer_norm` with `normalized_shape=(1,)`, the normalized slice has one element. In the standard LayerNorm formula, this forces the normalized activation to be zero:

```text
mean([x]) = x
var([x]) = 0
normalized = (x - mean) / sqrt(var + eps) = 0
output = bias
```

Therefore the output is independent of `weight`, and the gradient with respect to `weight` should be exactly zero. However, on CPU float32, PyTorch returns a nonzero `weight.grad`, even though the forward output is constant as `weight` changes.

I also see a related CPU bfloat16 forward issue for the same `normalized_shape=(1,)` case: PyTorch's saved stats report `mean == x`, but the output is not `bias`.

### Repro

```python
import torch
import torch.nn.functional as F

print("torch", torch.__version__)


def f(weight_value):
    x = torch.tensor([1e6], dtype=torch.float32)
    w = torch.tensor([weight_value], dtype=torch.float32)
    b = torch.tensor([3.0], dtype=torch.float32)
    return F.layer_norm(x, (1,), w, b, eps=1e-5)


print("f(w=0)   =", f(0.0).item())
print("f(w=2)   =", f(2.0).item())
print("f(w=123) =", f(123.0).item())

x = torch.tensor([1e6], dtype=torch.float32, requires_grad=True)
w = torch.tensor([2.0], dtype=torch.float32, requires_grad=True)
b = torch.tensor([3.0], dtype=torch.float32, requires_grad=True)

y = F.layer_norm(x, (1,), w, b, eps=1e-5)
y.backward()

print("y =", y.item())
print("x.grad =", x.grad.item())
print("w.grad =", w.grad.item())
print("b.grad =", b.grad.item())

out, mean, rstd = torch.ops.aten.native_layer_norm.default(
    x.detach(), [1], w.detach(), b.detach(), 1e-5
)
print("native_layer_norm out =", out.item())
print("native_layer_norm mean =", mean.item())
print("native_layer_norm rstd =", rstd.item())

# Related CPU bf16 forward check.
xb = torch.tensor([1e6], dtype=torch.bfloat16)
wb = torch.tensor([2.0], dtype=torch.bfloat16)
bb = torch.tensor([3.0], dtype=torch.bfloat16)
out_b, mean_b, rstd_b = torch.ops.aten.native_layer_norm.default(
    xb, [1], wb, bb, 1e-5
)
print("bf16 x stored =", float(xb))
print("bf16 out =", float(out_b))
print("bf16 mean =", float(mean_b))
print("bf16 rstd =", float(rstd_b))
```

### Actual output

On my machine with `torch 2.11.0+cu128`:

```text
f(w=0)   = 3.0
f(w=2)   = 3.0
f(w=123) = 3.0

y = 3.0
x.grad = 0.0
w.grad = 7.203125
b.grad = 1.0

native_layer_norm out = 3.0
native_layer_norm mean = 1000000.0
native_layer_norm rstd = 316.227783203125

bf16 x stored = 999424.0
bf16 out = 11.0
bf16 mean = 999424.0
bf16 rstd = 316.0
```

### Expected behavior

For the float32 backward case, `w.grad` should be exactly `0`. The forward output is independent of `weight`:

```text
f(w=0)   = 3.0
f(w=2)   = 3.0
f(w=123) = 3.0
```

so the nonzero `weight.grad` is not just a small numerical discrepancy; it contradicts the observed forward dependence.

For the bfloat16 forward case, `out` should be `bias == 3.0`, because PyTorch's own saved stats report `mean == x`, so `(x - mean) * rstd == 0`.

### Why this should hold

LayerNorm uses the standard formula:

```text
y = ((x - mean) / sqrt(var + eps)) * weight + bias
```

For a one-element normalized shape, `mean([x]) = x` and `var([x]) = 0`, so:

```text
normalized = 0
y = bias
dweight = sum(dy * normalized) = 0
```

This matches the LayerNorm paper and PyTorch's LayerNorm documentation:

- Layer Normalization, Ba/Kiros/Hinton 2016: https://arxiv.org/abs/1607.06450
- PyTorch LayerNorm docs: https://docs.pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html

### How this was found

We found this while comparing PyTorch normalization kernels against a Lean formalization in TorchLean and then running small property/fuzz-style tests over normalization edge cases.

TorchLean models normalization with the standard expression:

```text
y = ((x - mean) / sqrt(var + eps)) * weight + bias
```

and the scale-gradient contribution:

```text
dweight = sum(dy * normalized)
```

For a one-element normalized slice, the formal spec simplifies mechanically:

```text
mean = x
var = 0
normalized = 0
y = bias
dweight = 0
```

The core Lean-side contract we checked is essentially:

```lean
-- If a normalized slice is constant, TorchLean's normalization core returns the affine bias.
theorem constant_slice_normalizeCore_outputs_bias_check (x gamma beta epsilon : ℝ) :
    Spec.normalizeCore
        (s := .scalar)
        (s_mean := .scalar)
        (s_var := .scalar)
        (s_gamma := .scalar)
        (s_beta := .scalar)
        (epsilon := epsilon)
        (x := Spec.Tensor.scalar x)
        (mean := Spec.Tensor.scalar x)
        (variance := Spec.Tensor.scalar 0)
        (gamma := Spec.Tensor.scalar gamma)
        (beta := Spec.Tensor.scalar beta)
        (cb_mean := Spec.Shape.CanBroadcastTo.scalar_to_any .scalar)
        (cb_var := Spec.Shape.CanBroadcastTo.scalar_to_any .scalar)
        (cb_gamma := Spec.Shape.CanBroadcastTo.scalar_to_any .scalar)
        (cb_beta := Spec.Shape.CanBroadcastTo.scalar_to_any .scalar)
      = Spec.Tensor.scalar beta := by
  simp [Spec.normalizeCore, Spec.Tensor.broadcastTo, Spec.Tensor.addSpec, Spec.Tensor.subSpec,
    Spec.Tensor.mulSpec, Spec.Tensor.divSpec, Spec.Tensor.sqrtSpec, Spec.Tensor.mapSpec,
    Spec.Tensor.map2Spec, Spec.fill, Spec.replicate]
```

After seeing this proof obligation, we tested one-element normalized axes, constant slices, CPU/CUDA, float32/bfloat16, and native vs backend-dispatched paths. The repro above is the minimized failing LayerNorm case from that search.

### Possible source

This may be related to numerically unstable fused forms in the CPU LayerNorm kernel, e.g. computing an affine form equivalent to `a * x + b` with `b = -a * mean`, rather than preserving `(x - mean) * a` for this edge case.

Relevant source area:
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/layer_norm_kernel.cpp

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01
